File: ios/Classes/controller/RtpTransceiverController.swift
import Flutter

/// Controller of an `RtpTransceiver`.
class RtpTransceiverController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of the `RtpTransceiver`'s proxy.
  private var transceiver: RtpTransceiverProxy

  /// ID of the channel created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Initializes a new `RtpTransceiverController` for the provided
  /// `RtpTransceiverProxy`.
  init(messenger: FlutterBinaryMessenger, transceiver: RtpTransceiverProxy) {
    let channelName = ChannelNameGenerator.name(
      name: "RtpTransceiver",
      id: self.channelId
    )
    self.messenger = messenger
    self.transceiver = transceiver
    self.channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: messenger
    )
    self.channel.setMethodCallHandler { call, result in
      self.onMethodCall(call: call, result: result)
    }
  }

  /// Handles all the supported Flutter method calls for the controlled
  /// `RtpTransceiverProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "setDirection":
      let direction = argsMap!["direction"] as? Int
      self.transceiver
        .setDirection(direction: TransceiverDirection(rawValue: direction!)!)
      result(nil)
    case "setRecv":
      let enabled = argsMap!["recv"] as? Bool
      self.transceiver.setRecv(recv: enabled!)
      result(nil)
    case "setSend":
      let enabled = argsMap!["send"] as? Bool
      self.transceiver.setSend(send: enabled!)
      result(nil)
    case "getMid":
      let mid = self.transceiver.getMid()
      result(mid)
    case "getDirection":
      let direction = self.transceiver.getDirection()
      result(direction.rawValue)
    case "stop":
      self.transceiver.stop()
      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller into a Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": self.channelId,
      "sender": RtpSenderController(messenger: self.messenger,
                                    rtpSender: self.transceiver.getSender())
        .asFlutterResult(),
      "mid": self.transceiver.getMid(),
    ]
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/controller/RtpTransceiverController.kt
package com.instrumentisto.medea_flutter_webrtc.controller

import com.instrumentisto.medea_flutter_webrtc.model.CodecCapability
import com.instrumentisto.medea_flutter_webrtc.model.MediaType
import com.instrumentisto.medea_flutter_webrtc.model.RtpTransceiverDirection
import com.instrumentisto.medea_flutter_webrtc.proxy.RtpTransceiverProxy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Controller of [RtpTransceiverProxy] functional.
 *
 * @property messenger Messenger used for creating new [MethodChannel]s.
 * @property transceiver Underlying [RtpTransceiverProxy] to perform [MethodCall]s on.
 */
class RtpTransceiverController(
    private val messenger: BinaryMessenger,
    private val transceiver: RtpTransceiverProxy
) : MethodChannel.MethodCallHandler, IdentifiableController {
  /** Unique ID of the [MethodChannel] of this controller. */
  private val channelId = nextChannelId()

  /** Channel listened for the [MethodCall]s. */
  private val chan =
      MethodChannel(messenger, ChannelNameGenerator.name("RtpTransceiver", channelId))

  init {
    chan.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "setDirection" -> {
        val direction = RtpTransceiverDirection.fromInt(call.argument("direction")!!)
        transceiver.setDirection(direction)
        result.success(null)
      }
      "setCodecPreferences" -> {
        val args: List<Map<String, Any>> = call.argument("codecs") ?: listOf()
        val codecs =
            args.map {
              CodecCapability(
                  it["preferredPayloadType"] as Int,
                  it["name"] as String,
                  MediaType.fromInt(it["kind"] as Int),
                  it["clockRate"] as Int,
                  it["numChannels"] as Int?,
                  it["parameters"] as Map<String, String>,
                  it["mimeType"] as String)
            }
        transceiver.setCodecPreferences(codecs)
        result.success(null)
      }
      "setRecv" -> {
        val recv = call.argument<Boolean>("recv")!!
        transceiver.setRecv(recv)
        result.success(null)
      }
      "setSend" -> {
        val send = call.argument<Boolean>("send")!!
        transceiver.setSend(send)
        result.success(null)
      }
      "getMid" -> {
        result.success(transceiver.mid)
      }
      "getDirection" -> {
        result.success(transceiver.direction.value)
      }
      "stop" -> {
        transceiver.stop()
        result.success(null)
      }
      "dispose" -> {
        chan.setMethodCallHandler(null)
        result.success(null)
      }
    }
  }

  /**
   * Converts this [RtpTransceiverController] to the Flutter's method call result.
   *
   * @return [Map] generated from this controller which can be returned to the Flutter side.
   */
  fun asFlutterResult(): Map<String, Any?> {
    return mapOf(
        "channelId" to channelId,
        "sender" to RtpSenderController(messenger, transceiver.sender).asFlutterResult(),
        "mid" to transceiver.mid as Any?)
  }
}
